runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -73,96 +73,140 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
+        return readOrHoleFill(runtimeLayout, logResult, startAddress, numUnits, waitForWrite);
     }
 
     /**
-     * Reads a range of addresses from the Chain of log unit servers. This method optimizes for the time to wait to
-     * hole fill in case an empty address is encountered.
-     * If the waitForWrite flag is set to true, when an empty address is encountered, it waits for the hole to be
-     * filled. All subsequent empty addresses are hole filled directly and the reader does not wait.
-     * In case the flag is set to false, none of the reads wait for write completion and the empty address is hole
-     * filled right away.
+     * Reads a closed range of global addresses from the chain of log unit servers.
+     *
+     * - This method optimizes for the time to wait to hole fill in case an empty address
+     * is encountered.
+     * - If the waitForWrite flag is set to true, when an empty address is encountered,
+     * it waits for the hole to be filled. All subsequent empty addresses are hole filled
+     * directly and the reader does not wait.
+     * - In case the flag is set to false, none of the reads wait for write completion and
+     * the empty addresses are hole filled right away.
      *
-     * @param runtimeLayout Runtime layout.
-     * @param range         Range of addresses to read.
-     * @param waitForWrite  Flag whether wait for write is required or hole fill directly.
+     * @param runtimeLayout runtime layout.
+     * @param addresses     a closed range of addresses to read.
+     * @param waitForWrite  flag whether wait for write is required or hole fill directly.
      * @return Map of read addresses.
      */
-    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Range<Long> range, boolean waitForWrite) {
-        long startAddress = range.lowerEndpoint();
-        long endAddress = range.upperEndpoint();
+    @Override
+    @Nonnull
+    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout,
+                                         Range<Long> addresses,
+                                         boolean waitForWrite) {
+        long startAddress = addresses.lowerEndpoint();
+        long endAddress = addresses.upperEndpoint();
         int numUnits = runtimeLayout.getLayout().getSegmentLength(startAddress);
         log.trace("readRange[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
 
-        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(
-                runtimeLayout
-                        .getLogUnitClient(startAddress, numUnits - 1)
-                        .read(range)).getAddresses();
+        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(runtimeLayout
+                .getLogUnitClient(startAddress, numUnits - 1)
+                .readRange(addresses)).getAddresses();
+
+        return readOrHoleFill(runtimeLayout, logResult, startAddress, numUnits, waitForWrite);
```

Comment:
Factor out the fill hole logic, to take disjoint sets and ranges. Let's get rid of readOrHoleFill. It should be read Vs peak